In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

C:\Users\Admin\Downloads\anaconda\lib\site-packages\scipy\__init__.py:155: UserWarning: A NumPy version >=1.18.5 and <1.25.0 is required for this version of SciPy (detected version 1.26.1
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
df=pd.read_csv("prices.csv")

In [3]:
df

date symbol        open       close         low  \
0       2016-01-05 00:00:00   WLTW  123.430000  125.839996  122.309998   
1       2016-01-06 00:00:00   WLTW  125.239998  119.980003  119.940002   
2       2016-01-07 00:00:00   WLTW  116.379997  114.949997  114.930000   
3       2016-01-08 00:00:00   WLTW  115.480003  116.620003  113.500000   
4       2016-01-11 00:00:00   WLTW  117.010002  114.970001  114.089996   
...                     ...    ...         ...         ...         ...   
851259           2016-12-30    ZBH  103.309998  103.199997  102.849998   
851260           2016-12-30   ZION   43.070000   43.040001   42.689999   
851261           2016-12-30    ZTS   53.639999   53.529999   53.270000   
851262  2016-12-30 00:00:00    AIV   44.730000   45.450001   44.410000   
851263  2016-12-30 00:00:00    FTV   54.200001   53.630001   53.389999   

              high     volume  
0       126.250000  2163600.0  
1       125.540001  2386400.0  
2       119.739998  2489500.0  
3       117.440002  2006300.0  
4       117.330002  1408600.0  
...            ...        ...  
851259  103.930000   973800.0  
851260   43.310001  1938100.0  
851261   53.740002  1701200.0  
851262   45.590000  1380900.0  
851263   54.480000   705100.0  

[851264 rows x 7 columns]

In [4]:
df.drop(["date"],inplace=True,axis=1)

In [5]:
df

symbol        open       close         low        high     volume
0        WLTW  123.430000  125.839996  122.309998  126.250000  2163600.0
1        WLTW  125.239998  119.980003  119.940002  125.540001  2386400.0
2        WLTW  116.379997  114.949997  114.930000  119.739998  2489500.0
3        WLTW  115.480003  116.620003  113.500000  117.440002  2006300.0
4        WLTW  117.010002  114.970001  114.089996  117.330002  1408600.0
...       ...         ...         ...         ...         ...        ...
851259    ZBH  103.309998  103.199997  102.849998  103.930000   973800.0
851260   ZION   43.070000   43.040001   42.689999   43.310001  1938100.0
851261    ZTS   53.639999   53.529999   53.270000   53.740002  1701200.0
851262    AIV   44.730000   45.450001   44.410000   45.590000  1380900.0
851263    FTV   54.200001   53.630001   53.389999   54.480000   705100.0

[851264 rows x 6 columns]

In [6]:
df.drop(["symbol"],axis=1,inplace=True)

In [21]:
df.corr()

open     close       low      high    volume
open    1.000000  0.999849  0.999916  0.999930 -0.059950
close   0.999849  1.000000  0.999928  0.999927 -0.060154
low     0.999916  0.999928  1.000000  0.999892 -0.060760
high    0.999930  0.999927  0.999892  1.000000 -0.059452
volume -0.059950 -0.060154 -0.060760 -0.059452  1.000000

In [7]:
df.isnull().sum()

open      0
close     0
low       0
high      0
volume    0
dtype: int64

In [8]:
df.shape

(851264, 5)

In [9]:
from sklearn.model_selection import train_test_split

X = df.drop("close",axis=1)                                                           # MODEL BUILDING
y = df["close"]

X_train,X_test,Y_train,Y_test = train_test_split(X,y,test_size=0.20,random_state=0)

In [10]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train=sc.fit_transform(X_train)                                                      # Scaling
X_test=sc.fit_transform(X_test)

In [11]:
from sklearn.metrics import r2_score,mean_squared_error

In [12]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train,Y_train)                                                                   # TRAINING

Y_pred_lr = lr.predict(X_test)

In [13]:
score_lr = round(r2_score(Y_pred_lr,Y_test)*100,2)
mse=mean_squared_error(Y_pred_lr,Y_test)
print("The accuracy score achieved using Linear Regression is: "+str(score_lr)+ "%")     # Evaluation
print("The MSE achieved using Linear Regression is: "+str(mse))


The accuracy score achieved using Linear Regression is: 99.99%
The MSE achieved using Linear Regression is: 0.6446288377545947


In [14]:
from keras.models import Sequential
from keras.layers import Dense

In [15]:
model = Sequential()
model.add(Dense(64, input_dim=X_train.shape[1], activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(16, activation='relu'))
model.add(Dense(1, activation='linear'))


model.compile(loss='mean_squared_error', optimizer='adam')

In [16]:
model.fit(X_train, Y_train, epochs=50, batch_size=32, verbose=1)

Epoch 1/50
21282/21282 [==============================] - 87s 4ms/step - loss: 62.8267
Epoch 2/50
21282/21282 [==============================] - 84s 4ms/step - loss: 1.9720
Epoch 3/50
21282/21282 [==============================] - 85s 4ms/step - loss: 1.7394
Epoch 4/50
21282/21282 [==============================] - 86s 4ms/step - loss: 1.5703
Epoch 5/50
21282/21282 [==============================] - 85s 4ms/step - loss: 1.4374
Epoch 6/50
21282/21282 [==============================] - 84s 4ms/step - loss: 1.4059
Epoch 7/50
21282/21282 [==============================] - 84s 4ms/step - loss: 1.3351
Epoch 8/50
21282/21282 [==============================] - 84s 4ms/step - loss: 1.1926
Epoch 9/50
21282/21282 [==============================] - 83s 4ms/step - loss: 1.0974
Epoch 10/50
21282/21282 [==============================] - 84s 4ms/step - loss: 1.0942
Epoch 11/50
21282/21282 [==============================] - 86s 4ms/step - loss: 1.0488
Epoch 12/50
21282/21282 [==========================

In [17]:

Y_pred_ann = model.predict(X_test)

5321/5321 [==============================] - 15s 3ms/step


In [18]:
score_ann = round(r2_score(Y_pred_ann,Y_test)*100,2)

print("The accuracy score achieved using ANN is: "+str(score_lr)+" %")



The accuracy score achieved using ANN is: 99.99 %


In [19]:
mse=mean_squared_error(Y_pred_ann,Y_test)

In [20]:

print("The MSE achieved using ANN is: "+str(mse))

The MSE achieved using ANN is: 0.5836282552475494


# CONCLUSION:- AS YOU CAN SEE WE ACHIEVED 99.99% ACCURACY ON BOTH LINEAR AND ANN BUT AS WE CAN SEE THE AMOUNT OF LOSS HAS BEEN DECREASED ON ANN COMPARED TO CLASSIC LINEAR REGRESSION ALGORITHM